File: example-jni-java/src/RustJniGlueAdder.java
package com.maulingmonkey.jni_bindgen.example_jni_java;

public class RustJniGlueAdder implements Adder {
    @Override public native String add(String a, String b);
    @Override public native float add(float a, float b);
    @Override public native int add(int a, int b);

    static void test() {
        System.load(System.getProperty("com.maulingmonkey.jerk_test.jni_symbols_source")); // https://github.com/MaulingMonkey/jerk/issues/14
        Adder adder = new RustJniGlueAdder();
        assert adder.add("1", "2").equals("12");
        assert adder.add(1.0f, 2.0f) == 3.0f;
        assert adder.add(1, 2) == 3;
    }
}

File: example-jni-java/src/RustJniGlueAdder.rs
jni_glue::java! {
    import com.maulingmonkey.jni_bindgen.example_jni_java.RustJniGlueAdder;
    import com.maulingmonkey.jni_bindgen.example_jni_java.RustJniGlueAdder$Inner;
    import java.lang.String;

    unsafe impl class RustJniGlueAdder {
        float add(&env, this, float a, float b) {
            let r = a + b;
            Ok(r)
        }

        int add(&env, this, int a, int b) {
            let r = a + b;
            Ok(r)
        }

        String add(&env, this, String a, String b) {
            let r = format!("{}{}", a, b);
            println!("{} + {} = {}", a, b, r);
            Ok(r)
        }
    }
}

pub fn do_test() -> Result<(), jerk_test::JavaTestError> {
    jerk_test::run_test(
        "com.maulingmonkey.jni_bindgen.example_jni_java",
        "RustJniGlueAdder",
        "test"
    )
}

#[test] fn test() -> Result<(), jerk_test::JavaTestError> { do_test() }
